staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -412,12 +417,23 @@ func (w *watchCache) SetOnEvent(onEvent func(*watchCacheEvent)) {
 
 func (w *watchCache) GetAllEventsSinceThreadUnsafe(resourceVersion uint64) ([]*watchCacheEvent, error) {
 	size := w.endIndex - w.startIndex
-	// if we have no watch events in our cache, the oldest one we can successfully deliver to a watcher
-	// is the *next* event we'll receive, which will be at least one greater than our current resourceVersion
 	oldest := w.resourceVersion + 1
-	if size > 0 {
+	if size < w.capacity {
+		// If the watch cache isn't full it means that the oldest watch
+		// event we can deliver is at least one greater than the resource
+		// version of the last list.
+		if w.listResourceVersion != 0 {
+			oldest = w.listResourceVersion + 1
+		} else if size > 0 {
+			// If the list didn't happen, it's the first one in the buffer.
```

Comment:
Is this case even possible? If we hit this it seems like a bug

Reply:
Oh, I see, only happens in tests.

Reply:
Yeah - it's test-only thing. It shouldn't happen in production.

Reply:
But as I mentioned above - I think it's safer to add it here than to modify a bunch of other tests with this change.